utils/controller_debug.py

```diff
@@ -0,0 +1,789 @@
+    if ptype == "short":
+        dat = Split16(dat)
+    if ptype == "float":
+        dat = [F2I(x) for x in dat]
+        dat = Split32(dat)
+
+    SendCmd(OP_POKE, Split32(addr) + dat)
+
+
+def poke32(addr, dat):
+    poke(addr, dat, "long")
+
+
+def poke16(addr, dat):
+    poke(addr, dat, "short")
+
+
+def EscCmd(buff):
+    ret = []
+    for i in buff:
+        if i == ESC or i == TERM:
+            ret.append(ESC)
+        ret.append(i)
+    ret.append(TERM)
+    return ret
+
+
+def GetResp(show=False):
+    dat = []
+    esc = False
+    if show:
+        print("Getting resp: ", end="")
+    while True:
+        x = ser.read(1)
+        if len(x) < 1:
+            if show:
+                print("timeout")
+            return dat
+        x = ord(x)
+        if show:
+            print("0x%02x" % x, end=" ")
+
+        if esc:
+            esc = False
+            dat.append(x)
+            continue
+
+        if x == ESC:
+            esc = True
+            continue
+
+        if x == TERM:
+            if show:
+                print()
+            return dat
+        dat.append(x)
+
+
+def SendCmd(op, data=[], timeout=None):
+    global showSerial, ser
+    show = showSerial
+    buff = [op] + data
+
+    crc = crc16.calc(buff)
+    buff += Split16(crc)
+
+    if show:
+        S = "CMD: "
+        for b in buff:
+            S += "0x%02x " % b
+        print(S)
+
+    buff = EscCmd(buff)
+    if show:
+        S = "ESC: "
+        for b in buff:
+            S += "0x%02x " % b
+        print(S)
+
+    ser.write(bytearray(buff))
+    if timeout != None:
+        if show:
+            print("Setting timeout to %.1f" % timeout)
+        oldto = ser.timeout
+        ser.timeout = timeout
+
+    rsp = GetResp(show)
+    if timeout != None:
+        ser.timeout = oldto
+
+    if len(rsp) < 3:
+        raise Error("Invalid response, too short")
+
+    crc = crc16.calc(rsp[:-2])
+    rcrc = Build16(rsp[-2:])[0]
+    if crc != rcrc:
+        raise Error(
+            "CRC error on response, calculated 0x%04x received 0x%04x" % (crc, rcrc)
+        )
+
+    if rsp[0]:
+        raise Error("Error %d (0x%02x)" % (rsp[0], rsp[0]))
+        return []
+    return rsp[1:-2]
+
+
+def ReSync():
+    cmd = [TERM, TERM]
+    ser.write(bytearray(cmd))
+    time.sleep(0.1)
+    ser.reset_input_buffer()
+
+
+def I2F(ival):
+    s = struct.pack("I", ival & 0xFFFFFFFF)
+    return struct.unpack("f", s)[0]
+
+
+def F2I(fval):
+    s = struct.pack("f", fval)
+    return struct.unpack("I", s)[0]
+
+
+def SplitInt(val, len=4, le=True, asStr=False):
+    ret = []
+
+    if isinstance(val, list):
+        for v in val:
+            ret += SplitInt(v, len, le, False)
+    else:
+        for i in range(len):
+            ret.append(int(val & 0x00FF))
+            val >>= 8
+
+        if not le:
+            ret = ret[::-1]
+    if asStr:
+        ret = "".join([chr(i) for i in ret])
+    return ret
+
+
+def Split16(x, le=True, asStr=False):
```

Comment:
Can we use the struct library instead of the next bunch of Split() and Grab() functions?

Reply:
The grab functions are pretty handy for the way they're used.  The extract the first N bytes (depending on what's being grabbed) from the input list.  Not very efficient, but useful for where it's used here.